lib/registry.js

```diff
@@ -87,91 +88,122 @@ class RegistryFetcher extends Fetcher {
-              // made it this far, the integrity is worthwhile.  accept it.
-              // the setter here will take care of merging it into what we
-              // already had.
-              this.integrity = distIntegrity
+    const packument = await this.packument()
+    const mani = await pickManifest(packument, this.spec.fetchSpec, {
+      ...this.opts,
+      defaultTag: this.defaultTag,
+      before: this.before,
+    })
+    /* XXX add ETARGET and E403 revalidation of cached packuments here */
+
+    // add _resolved and _integrity from dist object
+    const { dist } = mani
+    if (dist) {
+      this.resolved = mani._resolved = dist.tarball
+      mani._from = this.from
+      const distIntegrity = dist.integrity ? ssri.parse(dist.integrity)
+        : dist.shasum ? ssri.fromHex(dist.shasum, 'sha1', { ...this.opts })
+        : null
+      if (distIntegrity) {
+        if (this.integrity && !this.integrity.match(distIntegrity)) {
+          // only bork if they have algos in common.
+          // otherwise we end up breaking if we have saved a sha512
+          // previously for the tarball, but the manifest only
+          // provides a sha1, which is possible for older publishes.
+          // Otherwise, this is almost certainly a case of holding it
+          // wrong, and will result in weird or insecure behavior
+          // later on when building package tree.
+          for (const algo of Object.keys(this.integrity)) {
+            if (distIntegrity[algo]) {
+              throw Object.assign(new Error(
+                `Integrity checksum failed when using ${algo}: ` +
+                `wanted ${this.integrity} but got ${distIntegrity}.`
+              ), { code: 'EINTEGRITY' })
             }
           }
         }
-        if (this.integrity) {
-          mani._integrity = String(this.integrity)
-          if (dist.signatures) {
+        // made it this far, the integrity is worthwhile.  accept it.
+        // the setter here will take care of merging it into what we already
+        // had.
+        this.integrity = distIntegrity
+      }
+    }
+    if (this.integrity) {
+      mani._integrity = String(this.integrity)
+      if (dist.signatures) {
+        if (this.opts.checkSignatures) {
+          if (this.registryKeys) {
+            // validate and throw on error, then set _signatures
+            const message = `${mani._id}:${mani._integrity}`
+            for (const signature of dist.signatures) {
+              const publicKey = this.registryKeys.filter(key => {
+                return (key.keyid === signature.keyid) &&
+                  !key.expires || (Date.parse(key.expires) > Date.now())
+              })[0]
+              if (!publicKey) {
+                throw Object.assign(new Error(
+                  'Integrity checksum signature failed: no valid public keys ' +
+                  `available for key ${signature.keyid}`
+                ), { code: 'EINTEGRITYSIGNATURE' })
+              }
```

Comment:
Thoughts on including the case for expired keys here and possibly updating the error message to something similar to what I noddled on in the [cli pr](https://github.com/npm/cli/pull/4827/files#diff-fb41d9148ee3f70333d72400c8e26026b419a2778140040b61d5591949f101bcR293-R304)?

I could also do those checks in CLI and then not expect this error case from happening but might get out of sync in future if we change logic in pacote and forget to update the CLI or vice versa.

Something like?
```suggestion
              const publicKey = this.registryKeys.filter(key => {
                return key.keyid === signature.keyid
              })[0]
              const validPublicKey = publicKey && (!publicKey.expires || (Date.parse(publicKey.expires) > Date.now())
              if (!publicKey && !validPublicKey) {
                throw Object.assign(new Error(
                  `${mani._id}` has a signature with keyid: ${signature.keyid} ` +
                  `but no corresponding public key can be found on ${this.registry}-/npm/v1/keys`
                ), { code: 'EMISSINGSIGNATUREKEY' })
              } else if (publicKey && !validPublicKey) {
                throw Object.assign(new Error(
                  `${mani._id} has a signature with keyid: ${signature.keyid} ` +
                  `but the corresponding public key on ${this.registry}-/npm/v1/keys has expired ` +
                  `(${publicKey.expires})`
                ), { code: 'EEXPIREDSIGNATUREKEY' })
              }
```

Also wondering if it would make sense to set a different error code for these? I was thinking to raise in the CLI if these two errors happen but [catch the error for invalid sigs](https://github.com/npm/cli/pull/4827/files#diff-fb41d9148ee3f70333d72400c8e26026b419a2778140040b61d5591949f101bcR312-R323) so I can add it to the list of invalid packages for the report.

Reply:
I don't think the cli should be checking the validity of the key.  The concern of the cli is parsing what the user wants and sending it to the right module, in this case pacote.  key validity is a "cryptographic" concern which is what pacote is doing here validating the signature.  We can add this check to pacote. I'll do this Monday.